mlir/lib/Dialect/EmitC/IR/FunctionOpAssembly.cpp

```diff
@@ -0,0 +1,335 @@
+      OpAsmParser::Delimiter::Paren, [&]() -> ParseResult {
+        // Ellipsis must be at end of the list.
+        if (isVariadic)
+          return parser.emitError(
+              parser.getCurrentLocation(),
+              "variadic arguments must be in the end of the argument list");
+
+        // Handle ellipsis as a special case.
+        if (allowVariadic && succeeded(parser.parseOptionalEllipsis())) {
+          // This is a variadic designator.
+          isVariadic = true;
+          return success(); // Stop parsing arguments.
+        }
+        // Parse argument name if present.
+        OpAsmParser::Argument argument;
+        auto argPresent = parser.parseOptionalArgument(
+            argument, /*allowType=*/true, /*allowAttrs=*/true);
+        if (argPresent.has_value()) {
+          if (failed(argPresent.value()))
+            return failure(); // Present but malformed.
+
+          // Reject this if the preceding argument was missing a name.
+          if (!arguments.empty() && arguments.back().ssaName.name.empty())
+            return parser.emitError(argument.ssaName.location,
+                                    "expected type instead of SSA identifier");
+          if (!parser.parseOptionalKeyword("ref")) {
+            llvm::ArrayRef<NamedAttribute> origAttrs;
+            if (!argument.attrs.empty())
+              origAttrs = argument.attrs.getValue();
+
+            SmallVector<NamedAttribute> attrs(origAttrs);
+            attrs.push_back(NamedAttribute(
+                StringAttr::get(parser.getContext(),
+                                emitc::getReferenceAttributeName()),
+                UnitAttr::get(parser.getContext())));
+            argument.attrs = DictionaryAttr::get(parser.getContext(), attrs);
+          }
+        } else {
+          argument.ssaName.location = parser.getCurrentLocation();
+          // Otherwise we just have a type list without SSA names.  Reject
+          // this if the preceding argument had a name.
+          if (!arguments.empty() && !arguments.back().ssaName.name.empty())
+            return parser.emitError(argument.ssaName.location,
+                                    "expected SSA identifier");
+
+          NamedAttrList attrs;
+          if (parser.parseType(argument.type) ||
+              parser.parseOptionalAttrDict(attrs) ||
+              parser.parseOptionalLocationSpecifier(argument.sourceLoc))
+            return failure();
+          if (!parser.parseOptionalKeyword("ref")) {
+            // Add attribute to argument
+            attrs.push_back(NamedAttribute(
+                StringAttr::get(parser.getContext(),
+                                emitc::getReferenceAttributeName()),
+                UnitAttr::get(parser.getContext())));
+          }
+          argument.attrs = attrs.getDictionary(parser.getContext());
+        }
+        arguments.push_back(argument);
+        return success();
+      });
+}
+
+/// Parse a function result list.
```

Comment:
I think `emitc.func` only allows for a single return value, it seems this will still parse a full list? Is this the exact same code as for `func.func`?